Upgrading yaml-rust to 0.4.2

gedigi opened the pull request:
Fixes #1387 

`cargo test --feature "yaml"` passes

gedigi commented:
@kbknapp could you please merge this and publish a new version of clap to crates.io so the security fix can be picked up? thanks!

BurntSushi commented:
@gedigi Why are you changing the map type being used? I would expect the commit message to mention this.

Also, the security issue seems inert in clap's case, since the yaml is written by the author. It's not used to decode end user input.

gedigi commented:
1- I changed the data structure because yaml-rust 0.4.* changed them too, and they have to be matched here - see #1396 

2- we cannot really predict how this library is used. it is possible that a user-controlled yaml file is passed to a setuid binary or a network service. also from a security compliance standpoint, having vulnerable libraries is something that needs to be justified on a case-by-case basis. I don't see the point of keeping an old, vulnerable library for something that is so widely used by clap when the change is so small

BurntSushi commented:
... I'm not arguing *against* this change.

huitseeker commented:
@kbknapp Any chance you'd have the time to have a quick look at this? It seems to be a straightforward upgrade which adds a [security fix](https://rustsec.org/advisories/RUSTSEC-2018-0006.html).
